### PR TITLE
:book: Add Encryption documentation

### LINKED ISSUE #624: 📖 documentation: kcrypt
Document how to use kcrypt

Depends on:

- https://github.com/kairos-io/kairos/issues/399
- https://github.com/kairos-io/kairos/issues/346

but we can start documenting what we have after 399 and add more as we go.

### PR DESCRIPTION
**Pull request by mudler**
Signed-off-by: mudler <mudler@c3os.io>

**What this PR does / why we need it**:
It adds encryption documentation, plus updates docsy and switches the syntax highlighter to `prism` so it does have a copy button to copy-paste the code blocks :partying_face: 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #624 

~~Still draft - this is a dump of my notes~~

### CODE REVIEW

File: docs/content/en/docs/Advanced/partition_encryption.md [hunk 1]
```diff
@@ -0,0 +1,250 @@
+---
+title: "Encrypting User Data with Kairos"
+linkTitle: "Encrypting User Data with Kairos"
+weight: 5
+description: >
+    Bundles are a powerful feature of Kairos that allow you to customize and configure your operating system. This section explains how to use and build custom bundles.
```
Comment (mudler) [inline docs/content/en/docs/Advanced/partition_encryption.md:6]: todo: change description

File: docs/content/en/docs/Advanced/partition_encryption.md [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+
+## Online mode 
+
+Online mode involves an external service (the Key Management Server, KMS) to boot the machines. The KMS role is to enable machine to boot by providing the encrypted secrets, or passphrases to unlock the encrypted drive. Authentication with the KMS is done via TPM challenging.
+
+In this scenario, we need to first deploy the KMS server to an existing Kubernetes cluster, and associate the TPM hash of the nodes that we want to manage. During deployment, we specify the KMS server inside the cloud-config of the nodes to be provisioned.
+
+### Requirements
+
+	- A Kubernetes cluster
```
Comment (mudler) [inline docs/content/en/docs/Advanced/partition_encryption.md:85]: nit: broken

File: docs/content/en/docs/Advanced/partition_encryption.md [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+  ssh_authorized_keys:
+  # Replace with your github user and un-comment the line below:
+  - github:mudler
+```
+
+## Troubleshooting  
+- Invoking `/system/discovery/kcrypt-discovery-challenger` without arguments returns the TPM pubhash.  
+- Invoking `kcrypt-discovery-challenger` with 'discovery.password' triggers the logic to retrieve the passphrase, for instance can be used as such:  
+	-  
+	  ```
```
Comment (mudler) [inline docs/content/en/docs/Advanced/partition_encryption.md:248]: nit: broken too

File: docs/content/en/docs/Advanced/partition_encryption.md [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+install:
+  encrypted_partitions:
+  - COS_PERSISTENT
+  grub_options:
+    extra_cmdline: "rd.neednet=1"
+
+kcrypt:
+  challenger:
+    challenger_server: "http://192.168.68.109:30000"
+    nv_index: ""
```
Comment (mudler) [inline docs/content/en/docs/Advanced/partition_encryption.md:230]: todo: need to document those

File: docs/content/en/docs/Advanced/partition_encryption.md [hunk 5]
```diff
@@ -0,0 +1,286 @@
+---
+title: "Encrypting User Data with Kairos"
+linkTitle: "Encrypting User Data with Kairos"
+weight: 5
+description: >
+    This section describes how to encrypt partition with LUKS in Kairos.
+---
+
+{{% alert title="Note" color="warning" %}}
```
Comment (Itxaka) [inline docs/content/en/docs/Advanced/partition_encryption.md:9]: MAybe we should have a code block to mark when X feature was available from, like in goreleaser docs, I really like that stuff and its reusable:

image

File: docs/content/en/docs/Advanced/partition_encryption.md [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+| --- | --- |
+| `install.encrypted_partitions` | Label of partitions to encrypt |
+| `kcrypt.challenger.challenger_server` | External KMS Server address |
+| `kcrypt.challenger.nv_index` | Custom Non-Volatile index to use to store encoded blobs |
+| `kcrypt.challenger.c_index` | Custom Index for the RSA Key pair |
+| `kcrypt.challenger.tpm_device` | Custom TPM device |
+
+## Requirements
+
+The host machine must have a TPM chip version 2.0 or higher to use encryption with Kairos. A list of TPM chips/HW can be found [in the list of certified products](https://trustedcomputinggroup.org/membership/certification/tpm-certified-products/), however, any modern machine have a TPM 2.0 chip.
```
Comment (Itxaka) [inline docs/content/en/docs/Advanced/partition_encryption.md:75]:
```suggestion
The host machine must have a TPM chip version 2.0 or higher to use encryption with Kairos. A list of TPM chips/HW can be found [in the list of certified products](https://trustedcomputinggroup.org/membership/certification/tpm-certified-products/), however, any modern machine has a TPM 2.0 chip.
```

File: docs/content/en/docs/Advanced/partition_encryption.md [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+users:
+- name: kairos
+  # Change to your pass here
+  passwd: kairos
+  ssh_authorized_keys:
+  # Replace with your github user and un-comment the line below:
+  # - github:mudler
+```
+
+Note, we define a list of partition labels that we want to encrypt. In the examble above we set `COS_PERSISTENT` to be encrypted, which in turns will encrypt all the user-data of the machine (this includes, for instance, Kubernetes pulled images, or any runtime persisting data on the machine).
```
Comment (Itxaka) [inline docs/content/en/docs/Advanced/partition_encryption.md:114]:
```suggestion
Note, we define a list of partition labels that we want to encrypt. In the example above we set `COS_PERSISTENT` to be encrypted, which in turns will encrypt all the user-data of the machine (this includes, for instance, Kubernetes pulled images, or any runtime persisting data on the machine).
```

File: docs/content/en/docs/Getting started/_index.md [hunk 8]
```diff
@@ -24,16 +24,17 @@ Once the installation is complete, you can begin using your Kubernetes cluster.
...
 
 {{% alert title="Note" %}}
-The releases in the [kairos-io/kairos](https://github.com/kairos-io/kairos/releases) repository are the Kairos core images that ship **without** K3s and P2P full-mesh functionalities; however, further extensions can be installed dynamically in runtime by using the Kairos bundles mechanism.
+The releases in the [kairos-io/kairos](https://github.com/kairos-io/kairos/releases) repository are the Kairos core images that ship **without** K3s and P2P full-mesh functionalities; Core images can be used as a generic installer to [deploy container images](/docs/examples/core).
 
-The releases in [kairos-io/provider-kairos](https://github.com/kairos-io/provider-kairos/releases) ship **with** k3s and P2P full-mesh instead. These options need to be explicitly enabled. In follow-up releases, _k3s-only_ artifacts will also be available.
+The releases in [kairos-io/provider-kairos](https://github.com/kairos-io/provider-kairos/releases) contains already **with** k3s and P2P full-mesh instead. These options need to be explicitly enabled. In follow-up releases, _k3s-only_ artifacts will also be available.
```
Comment (Itxaka) [inline docs/content/en/docs/Getting started/_index.md:37]: `contains already with k3s` sounds pretty bad. Cannot put my finger on why but it does sound strange to me. `comes with k3s and...` or  `bundles k3s and...` instead maybe?

File: docs/content/en/docs/Advanced/partition_encryption.md [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+This feature will be available in Kairos version `1.5.0` and in all future releases.
+
+{{% /alert %}}
+
+Kairos offers the ability to encrypt user data partitions with `LUKS`. User-data partitions are dedicated to persist data for a running system, stored separately from the OS images. This encryption mechanism can also be used to encrypt additional partitions created during the installation process.  
+
+Kairos supports the following encryption scenarios:  
+
+1. **Offline mode** - Encryption key for partitions is stored on the machine inside the TPM chip.  
+1. **Online mode (Automated)** - Keypair used to encrypt the partition passphrase is stored on the TPM chip.
```
Comment (Itxaka) [inline docs/content/en/docs/Advanced/partition_encryption.md:20]: In this short view, I cant discern between the ofline and online(automated) modes easily. Maybe a more precise description or and advantage or something? 

Its clear to me comparing to the online(manual) config at a glance what it entails, not so much with the other 2.

Comment (mudler) [reply docs/content/en/docs/Advanced/partition_encryption.md]: good point! changed in https://github.com/kairos-io/kairos/pull/722/commits/f56a73669d480298f9f73c6f6db5608855264128

File: docs/content/en/docs/Getting started/_index.md (hunk 8 above)
Comment (mudler) [reply docs/content/en/docs/Getting started/_index.md]: rephrased in https://github.com/kairos-io/kairos/pull/722/commits/f56a73669d480298f9f73c6f6db5608855264128

File: docs/content/en/docs/Advanced/partition_encryption.md [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+    tpm_device: ""
+
+hostname: metal-{{ trunc 4 .MachineID }}
+users:
+- name: kairos
+  # Change to your pass here
+  passwd: kairos
+  ssh_authorized_keys:
+  # Replace with your github user and un-comment the line below:
+  - github:mudler
```
Comment (mudler) [inline docs/content/en/docs/Advanced/partition_encryption.md:278]:
```suggestion
  # - github:mudler
```

File: docs/content/en/docs/Advanced/partition_encryption.md [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
```
Comment (mudler) [inline docs/content/en/docs/Advanced/partition_encryption.md:217]:
```suggestion
  # - github:mudler
```

File: docs/content/en/docs/Advanced/partition_encryption.md [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+This feature will be available in Kairos version `1.5.0` and in all future releases.
+
+{{% /alert %}}
+
+Kairos offers the ability to encrypt user data partitions with `LUKS`. User-data partitions are dedicated to persist data for a running system, stored separately from the OS images. This encryption mechanism can also be used to encrypt additional partitions created during the installation process.  
+
+Kairos supports the following encryption scenarios:  
+
+1. **Offline mode** - Encryption key for partitions is stored on the machine inside the TPM chip.  
+1. **Online mode (Automated)** - Keypair used to encrypt the partition passphrase is stored on the TPM chip, and an external server is used to store the encrypted passphrases.
```
Comment (Itxaka) [inline docs/content/en/docs/Advanced/partition_encryption.md:20]: nice, much clearer now!

**Review by Itxaka — APPROVED** [review]

**Comment by jimmykarily** [comment]
Awesome work!

**Pull request merged**